torch/distributions/truncated_distribution.py

```diff
@@ -0,0 +1,61 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all
+
+
+class TruncatedDistribution(Distribution):
+    r"""
+    Extension of the Distribution class, which applies truncation to base distribution.
+    """
+    def __init__(self, base_distribution, lower_bound=-float('inf'), upper_bound=float('inf'), *args, **kwargs):
+        super(TruncatedDistribution, self).__init__(*args, **kwargs)
+        self.base_dist = base_distribution
+        self.lower_bound, self.upper_bound, _ = broadcast_all(lower_bound, upper_bound,
+                                                              getattr(self.base_dist,
+                                                                      list(self.base_dist.params.keys())[0]))
+
+    @constraints.dependent_property
+    def params(self):
+        return self.base_dist.params  # TODO add params of truncation
+
+    @constraints.dependent_property
+    def support(self):
+        # Note: The proper way to implement this is intersection([lower_bound, upper_bound], base_dist.support)
+        # This requires intersection method to be implemented for constraints.
+        return constraints.interval(self.lower_bound, self.upper_bound)
+
+    @property
+    def batch_shape(self):
+        return self.base_dist.batch_shape
+
+    @property
+    def event_shape(self):
+        return self.base_dist.event_shape
+
+    def rsample(self, sample_shape=torch.Size()):
+        """
+        Generates a sample_shape shaped sample or sample_shape shaped batch of
+        samples if the distribution parameters are batched via inverse cdf sampling. Note that this
+        is a generic sampler which is not the most efficient or accurate around tails of base distribution.
+        """
+        shape = shape = self._extended_shape(sample_shape)
+        u = getattr(self.base_dist, list(self.base_dist.params.keys())[0]).new(shape).uniform_()
```

Comment:
This looks dangerous. I wish we had a `.new()` method to create a correctly-placed tensor from given distribution.

@apaszke Is there an established pattern to do this? Can we define a `.new_tensor()` method or something? This has been coming up often. Some of our distributions define a private `_new()` but we haven't exposed this as a general interface.

Reply:
It seems simple and safe to define a method-as-property like
```py
class Distribution(object):
    @property
    def new_tensor(self):
        raise NotImplementedError

class Normal(Distribution):
    @property
    def new_tensor(self):
        return self.loc.new
```